crates/bdk/src/wallet/mod.rs

```diff
@@ -310,7 +358,9 @@ impl<D> Wallet<D> {
         };
 
         let changeset = db.load_from_persistence().map_err(NewError::Persist)?;
-        chain.apply_changeset(&changeset.chain);
+        chain
+            .apply_changeset(&changeset.chain)
+            .map_err(|_| NewError::InvalidPersistenceGenesis)?;
```

Comment:
I think we should consider a explicit API  which distinguishes between initializing a wallet for the first time and loading it -- so `::new` and `::load`. It should be simpler for developers and could avoid this error case. Furthermore consider the implications of also doing this: https://github.com/bitcoindevkit/bdk/issues/1101#issuecomment-1701929749 and we also persist the `Network` in the first changeset (waving hands a it here). Then load could load everything from in one go without the possibility of contradiction between runtime data and persisted data.

So `new` would take what it takes now but would persist the genesis block, network and keychains immediately. `load` would get everything from persistence and applications would just inspect it at runtime to find out what kind of wallet the user had. If we really want this kind of function we could have `new_or_load` as well which just tries to load, checks everything is consistent with the paramters (throwing errors like above if not) or if it can't be loaded, making a new one.

Reply:
I think `Wallet::new` probably needs a check as to whether the underlying `PersistenceBackend` is empty. Otherwise, there may be accidental overwrites.

Reply:
I don't think the `PersistenceBackend` has an overwrite API though?

Reply:
When `bdk_file_store::Store` is initialized, the logical cursor is positioned after the "magic bytes".

`Wallet::new` does not load any data from the `PersistenceBackend`.

So the next write will start at the position after "magic bytes" (if I'm not mistaken).

Reply:
So this sounds like an implementation specific problem. I don't think that `Wallet::new` needs to check anything here. We can just have a safe method in `bdk_file_store` that errors out when you create a new db for a file that already exists. Remember the data we are storing is fine to delete. It can always be reconstructed.

Reply:
I now have separate methods on `bdk_file_store::Store` for initializing and loading.

Reply:
So I didn't get feedback on the idea of doing #1101 as part of this. Is this PR intended to be a stepping stone towards doing #1101 or do you not think #1101 is viable. I doesn't seem valuable to do this change and then change the API again when we do #1101.

Reply:
How about we first move the descriptors to a field in the wallet's ChangeSet. Then in a separate PR, we change the keychain txout index to have changesets with descriptors? This way the API of wallet stays consistent.